core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -605,56 +613,73 @@ public SourceDistributionTaskSource(
+        {
+            ImmutableList.Builder<TaskDescriptor> result = ImmutableList.builder();
+            while (true) {
+                Optional<TaskDescriptor> readyTask = getReadyTask(splits, otherSplitSets, nodeRequirements);
+                if (readyTask.isEmpty()) {
+                    break;
+                }
+                result.add(readyTask.get());
             }
 
-            if (!locallyAccessibleSplitBuffer.isEmpty()) {
-                for (HostAddress remoteHost : locallyAccessibleSplitBuffer.keySet()) {
-                    List<Split> splits = ImmutableList.copyOf(locallyAccessibleSplitBuffer.get(remoteHost));
-                    if (!splits.isEmpty()) {
-                        locallyAccessibleSplitBuffer.values().forEach(otherHostSplits -> splits.forEach(otherHostSplits::remove));
-                        result.add(new TaskDescriptor(
-                                currentPartitionId++,
-                                ImmutableListMultimap.<PlanNodeId, Split>builder().putAll(partitionedSourceNodeId, splits).build(),
-                                replicatedExchangeSourceHandles,
-                                new NodeRequirements(catalogRequirement, ImmutableSet.of(remoteHost))));
-                    }
+            if (includeRemainder && !splits.isEmpty()) {
+                result.add(buildTaskDescriptor(splits, nodeRequirements));
+                for (Set<Split> otherSplits : otherSplitSets) {
+                    otherSplits.removeAll(splits);
                 }
-                locallyAccessibleSplitBuffer.clear();
+                splits.removeAll(splits);
             }
-
             return result.build();
         }
 
-        private static <T> List<T> removeN(Collection<T> collection, int n)
+        private Optional<TaskDescriptor> getReadyTask(Set<Split> splits, List<Set<Split>> otherSplitSets, NodeRequirements nodeRequirements)
         {
-            ImmutableList.Builder<T> result = ImmutableList.builder();
-            Iterator<T> iterator = collection.iterator();
-            for (int i = 0; i < n && iterator.hasNext(); i++) {
-                T item = iterator.next();
-                iterator.remove();
-                result.add(item);
+            ImmutableList.Builder<Split> chosenSplitsBuilder = ImmutableList.builder();
+            int splitCount = 0;
+            int totalSlitWeight = 0;
+            for (Split split : splits) {
+                totalSlitWeight += split.getSplitWeight().getRawValue();
+                splitCount++;
+                chosenSplitsBuilder.add(split);
+
+                if (totalSlitWeight >= targetPartitionSplitWeight && splitCount >= minPartitionSplitCount || splitCount >= maxPartitionSplitCount) {
+                    ImmutableList<Split> chosenSplits = chosenSplitsBuilder.build();
+                    for (Set<Split> otherSplits : otherSplitSets) {
+                        otherSplits.removeAll(chosenSplits);
```

Comment:
For some reason Idea doesn't like `removeAll` and suggests that it might work "slowly" with an advice to replace it with `chosenSplits.forEach(otherSplits::remove)`. I remember it was something related to `removeAll` using `contains` that might be slow in some cases. Maybe it's not the case here.

Reply:
replaced just in case